### PR TITLE
Consolidated search improvements

### PR DESCRIPTION
**Pull request by maartenba**
Apologies for making this a big commit. This has been an ongoing effort where all of the below sort of link together and were very hard to commit separately.

**Technical changes**

* Introduce cloud service (instead of using Antares)
* Install IIS AppInit on the web role and make sure the index is warm before the machine gets into load balancer

**Performance improvements**

* Remove all user code lock statements where possible
* AzureDirectory is used as source store, which is synced with an in-memory index every 5 minutes (no more waits for locks from AzureDirectory)
* Index load time on cold start ~2 minutes, warm start (with a cache on disk) ~1 minute
* Rewrite V2 "id:xyz" queries that originate from the NuGet Gallery to "packageid:xyz" to increase matching performance
* Rewrite ranking result scoring to use a map of Lucene document id's (array index lookup vs. dictionary find)
* Rewrite package owners to be merged in search results based on a simple dictionary instead of indexing them on the fly (reduced startup time by ~5 minutes)
* Rewrite queries on package owners to use a filter (reduces the document result set before running tehactual query, improving query performance)
* String interning for package id's and package versions to reduce GC overhead (also reduced managed memory footprint by ~100 MB)
* After the above changes, the response cache we introduced earlier became the bottleneck - so I removed it

**Functional improvements**

* V3 compatibility (don't include unlisted packages in result by default)
* V2 compatibility (don't filter listed/unlisted and prerelease as current V2 also does not do this)

**Remarks**

* The unit tests are still failing (intentional) - I have to redo these as the shape of our queries has changed by this commit. Want to first gather feedback on the PR before fixing the tests and re-doing the tests.
* Performance profiling on an Azure D3v2 machine shows that this code can take 1.700 requests per second on one machine. Current V2 search handles ~500/sec on average, on 6 machines.
* I expect us to have to run on 4 machines in cloud services for redundancy + ensuring spare capacity.

@skofman1 @xavierdecoster @joelverhagen @johnataylor Care to review? :-)

### CODE REVIEW

File: src/NuGet.Services.BasicSearch/WebRole.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                {
+                    using (var reader = new StreamReader(response.GetResponseStream()))
+                    {
+                        reader.ReadToEnd();
+                    }
+                }
+            }
+            catch
+            {
+                // ignore
```
Comment (skofman1) [inline src/NuGet.Services.BasicSearch/WebRole.cs]: log?

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 2]
```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
             var searcher = searcherManager.Get();
             try
             {
-                bool includeUnlisted = ignoreFilter;
-                includePrerelease = ignoreFilter || includePrerelease;
+                // The old V2 search service would tread "id:" queries (~match) in the same way as it did "packageid:" (==match).
```
Comment (skofman1) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: treat

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 3]
```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
...
+                // If "id:" is in the query, replace it.
+                if (luceneQuery && !string.IsNullOrEmpty(q) && q.StartsWith("id:", StringComparison.OrdinalIgnoreCase))
+                {
+                    q = "packageid:" + q.Substring(3);
```
Comment (skofman1) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: any reason not to use string.Replace?

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 4]
```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
...
+                // the data you'd expect looking at the method parameters:
+                //   bool includeUnlisted = ignoreFilter;
+                //   includePrerelease = ignoreFilter || includePrerelease;
+                //   feed = ignoreFilter ? null : feed;
+                //   Filter filter = searcher.GetFilter(includeUnlisted, includePrerelease, feed);
+                //
+                // But hey, we'll do it the old V2 way instead and happily ignore the correct way.
                 feed = ignoreFilter ? null : feed;
-                Filter filter = searcher.GetFilter(includeUnlisted, includePrerelease, feed);
+                Filter filter = searcher.GetFilter(feed);
```
Comment (skofman1) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: it confusing that you pass a null to GetFilter, and expect it to return a correct value. A clearer approach would be:
Filter filter = ignoreFilter ? null : searcher.GetFilter(feed);

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 5]
```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
...
 
-                Query query = NuGetQuery.MakeQuery(q);
+                Query query = NuGetQuery.MakeQuery(q, searcher);
+
+                if (filter != null)
+                {
+                    query = new FilteredQuery(query, filter);
+                    filter = null;
```
Comment (skofman1) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: why?

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 6]
```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
...
+                }
 
                 if (countOnly)
                 {
                     DocumentCountImpl(jsonWriter, searcher, query, filter);
```
Comment (skofman1) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: you always pass null here, so just pass null explicitly

File: src/NuGet.Indexing/GalleryServiceImpl.cs [hunk 7]
```diff
@@ -25,22 +25,49 @@ public static void Search(JsonWriter jsonWriter, NuGetSearcherManager searcherMa
...
                 }
                 else
                 {
-                    IDictionary<string, int> rankings = searcher.Rankings;
-
-                    ListDocumentsImpl(jsonWriter, searcher, query, rankings, filter, sortBy, skip, take, searcherManager);
+                    ListDocumentsImpl(jsonWriter, searcher, query, filter, sortBy, skip, take, searcherManager);
```
Comment (skofman1) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: same, just pass null explicitly

File: src/NuGet.Services.BasicSearch/WebRole.cs (hunk 1 above)
Comment (maartenba) [inline src/NuGet.Services.BasicSearch/WebRole.cs]: not important here, it's mainly to do a first request to the local host when just started

File: src/NuGet.Indexing/RankingsScoreQuery.cs [hunk 8]
```diff
@@ -39,41 +38,54 @@ public override string ToString()
...
             {
                 _rankings = rankings;
-                _ids = FieldCache_Fields.DEFAULT.GetStrings(reader, "Id");
                 _baseBoost = baseBoost;
+
+                SegmentReader segmentReader = reader as SegmentReader;
+
+                _readerName = segmentReader != null
+                    ? segmentReader.SegmentName
+                    : string.Empty;
```
Comment (skofman1) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: this is unclear. Can you please add a comment explaining the behavior here?

Comment (skofman1) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: And in general, can you add some documentation to this file..

---
In reply to: [59756131](https://github.com/NuGet/NuGet.Services.Metadata/pull/64#discussion_r59756131) [](ancestors = 59756131)

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 2 above)
Comment (maartenba) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Fixed

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 4 above)
Comment (maartenba) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Agreed. Updated.

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 5 above)
Comment (maartenba) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: This filters before running the query, reducing the search set (and the amount of results to search a score/ranking for). Adding a comment.

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 6 above)
Comment (maartenba) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Fixed, thanks!

File: src/NuGet.Indexing/GalleryServiceImpl.cs (hunk 7 above)
Comment (maartenba) [inline src/NuGet.Indexing/GalleryServiceImpl.cs]: Fixed

File: src/NuGet.Indexing/RankingsScoreQuery.cs (hunk 8 above)
Comment (maartenba) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: Would it make sense to do a quick meeting around the entire thing?

Comment (maartenba) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: Added some comments, how does it look?

File: src/NuGet.Indexing/NuGetSearcherManager.cs [hunk 9]
```diff
@@ -102,11 +115,21 @@ public class NuGetSearcherManager : SearcherManager<NuGetIndexSearcher>
...
+
+                    azureDirectorySynchronizer = new AzureDirectorySynchronizer(sourceDirectory, directory);
+                }
+
                 searcherManager = new NuGetSearcherManager(
                     indexContainer,
                     loggerFactory.CreateLogger<NuGetSearcherManager>(),
-                    directory ?? new AzureDirectory(storageAccount, indexContainer),
-                    loader ?? new StorageLoader(storageAccount, dataContainer, loggerFactory.CreateLogger<StorageLoader>()));
+                    directory ?? new AzureDirectory(storageAccount, indexContainer, new RAMDirectory()),
```
Comment (skofman1) [inline src/NuGet.Indexing/NuGetSearcherManager.cs]: directory always has value at this point

File: src/NuGet.Indexing/NuGetSearcherManager.cs [hunk 10]
```diff
@@ -117,20 +140,37 @@ public class NuGetSearcherManager : SearcherManager<NuGetIndexSearcher>
...
 
             _logger.LogInformation("NuGetSearcherManager.Reopen: refreshed original IndexReader in {IndexReaderReopenDuration} seconds.", stopwatch.Elapsed.TotalSeconds);
 
             return indexReader;
         }
 
+        protected override bool RequiresNewSearcher(IndexReader newReader, IndexSearcher currentSearcher)
+        {
+            return newReader != currentSearcher.IndexReader;
+        }
```
Comment (skofman1) [inline src/NuGet.Indexing/NuGetSearcherManager.cs]: it's the same implementation everywhere, so why do u need to override?

File: src/NuGet.Indexing/NuGetSearcherManager.cs (hunk 9 above)
Comment (maartenba) [inline src/NuGet.Indexing/NuGetSearcherManager.cs]: 🙇

File: src/NuGet.Indexing/NuGetSearcherManager.cs (hunk 10 above)
Comment (maartenba) [inline src/NuGet.Indexing/NuGetSearcherManager.cs]: Fixed

File: src/NuGet.Indexing/NuGetSearcherManager.cs [hunk 11]
```diff
@@ -200,48 +220,68 @@ protected override NuGetIndexSearcher CreateSearcher(IndexReader reader)
...
-                    rankings,
+                    rankingsHandler.Result,
                     latestBitSet,
-                    latestStableBitSet);
+                    latestStableBitSet,
+                    ownersHandler.Result);
             }
             catch (Exception e)
             {
                 _logger.LogError("Internal server error", e);
```
Comment (skofman1) [inline src/NuGet.Indexing/NuGetSearcherManager.cs]: more details like method name, etc.

File: src/NuGet.Indexing/NuGetIndexSearcher.cs [hunk 12]
```diff
@@ -12,43 +13,59 @@ namespace NuGet.Indexing
...
-        public IDictionary<string, int> Rankings { get; private set; }
+        public RankingsHandler.RankingResult Rankings { get; private set; }
         public OpenBitSet LatestBitSet { get; private set; }
         public OpenBitSet LatestStableBitSet { get; private set; }
+        public OwnersHandler.OwnersResult Owners { get; private set; }
         public DateTime LastReopen { get; private set; }
 
+        public Filter GetFilter(string curatedFeed)
+        {
+            Filter curatedFeedFilter;
```
Comment (skofman1) [inline src/NuGet.Indexing/NuGetIndexSearcher.cs]: this should throw id curatedFeed is empty

Comment (skofman1) [inline src/NuGet.Indexing/NuGetIndexSearcher.cs]: Or change it to:
bool TryGetFilter(string curatedFeed, out Filter filter)

---
In reply to: [59766156](https://github.com/NuGet/NuGet.Services.Metadata/pull/64#discussion_r59766156) [](ancestors = 59766156)

File: src/NuGet.Indexing/AzureDirectorySynchronizer.cs [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+                    }
+                    finally
+                    {
+                        try
+                        {
+                            indexOutput?.Close();
+                        }
+                        finally
+                        {
+                            indexInput?.Close();
```
Comment (skofman1) [inline src/NuGet.Indexing/AzureDirectorySynchronizer.cs]: can this throw? if so wrap it in try catch

File: src/NuGet.Indexing/CuratedFeedHandler.cs [hunk 14]
```diff
@@ -59,13 +59,13 @@ public void End(IndexReader indexReader)
             }
         }
 
-        public void Process(IndexReader indexReader, string readerName, int n, Document document, string id, NuGetVersion version)
+        public void Process(IndexReader indexReader, string readerName, int documentNumber, Document document, string id, NuGetVersion version)
```
Comment (skofman1) [inline src/NuGet.Indexing/CuratedFeedHandler.cs]: good one :)

File: src/NuGet.Indexing/RankingsScoreQuery.cs [hunk 15]
```diff
@@ -39,41 +39,59 @@ public override string ToString()
...
 
-            public RankingScoreProvider(IndexReader reader, IDictionary<string, int> rankings, double baseBoost)
+            public RankingScoreProvider(IndexReader reader, RankingsHandler.RankingResult rankings, double baseBoost)
                 : base(reader)
             {
                 _rankings = rankings;
-                _ids = FieldCache_Fields.DEFAULT.GetStrings(reader, "Id");
                 _baseBoost = baseBoost;
+
+                // We need the reader name: Lucene *may* has multiple segments (which are smaller indices)
```
Comment (skofman1) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: have

**Comment by skofman1** [comment]
LGTM

File: src/NuGet.Indexing/NuGetSearcherManager.cs (hunk 11 above)
Comment (maartenba) [inline src/NuGet.Indexing/NuGetSearcherManager.cs]: Fixed, also added some more granularity in things that could go wrong in that method.

File: src/NuGet.Indexing/NuGetIndexSearcher.cs (hunk 12 above)
Comment (maartenba) [inline src/NuGet.Indexing/NuGetIndexSearcher.cs]: Updated.

File: src/NuGet.Indexing/AzureDirectorySynchronizer.cs (hunk 13 above)
Comment (maartenba) [inline src/NuGet.Indexing/AzureDirectorySynchronizer.cs]: Caught in NuGetSearcherManager - IndexReader Reopen(IndexSearcher searcher)

File: src/NuGet.Indexing/CuratedFeedHandler.cs (hunk 14 above)
Comment (maartenba) [inline src/NuGet.Indexing/CuratedFeedHandler.cs]: Clarity for the win :-)

File: src/NuGet.Indexing/RankingsScoreQuery.cs (hunk 15 above)
Comment (maartenba) [inline src/NuGet.Indexing/RankingsScoreQuery.cs]: Fixed, thanks!

**Pull request merged**